### PR TITLE
Migrate global to script scope for variables

### PR DESCRIPTION
**Pull request by FabienTschanz**
**Please note: This PR has a dependency on #185**

This PR changes how `MSCloudLoginAssistant` stores its connection information for the different workloads. Previously, it used `$global` scoped variables, but those are not available between resources when running any `*-DscConfiguration` cmdlet. The scope would be reset between resources, forcing a reauthentication for every resource processed. See https://github.com/microsoft/Microsoft365DSC/issues/4982 for more information.

To achieve this functionality, the module now doesn't depend anymore on "submodules" where a common state would not have been possible to share and persist. The submodules were replaced by *.ps1 (PowerShell script) instead of *.psm1 (PowerShell module). This allows to dot-source functions into the current running module scope and all code inside of it has access to the same `$script` scoped variables.

Additionally, the log messages were all rewritten to use `Add-MSCloudLoginAssistantEvent`, similar to `Add-M365DSCEvent` - A function that writes to an event log if the environment variable `MSCLOUDLOGINASSISTANT_WRITETOEVENTLOG` is set to `true`. This way, the console log is not being clustered with too much information, but rather stays clean in its own little world.

- - -
This change is [Reviewable](https://reviewable.io/reviews/microsoft/MSCloudLoginAssistant/188)

### CODE REVIEW

**Comment by Borgquite** [comment]
Here's confirmation that this is a good way to solve it (based on others'e experience of the same issue of trying to share a token between commands in a module to avoid authentication taking place for each command)

Also seems to confirm that sharing variables within commands using script scope and separate dot-sourced PS1 files seems to be a valid approach.

https://mikefrobbins.com/2017/06/08/what-is-this-module-scope-in-powershell-that-you-speak-of/

**Comment by FabienTschanz** [comment]
@Borgquite Indeed. I have some other private modules that use the same approach, with much success so far (also for an authentication token shared across commands).

File: Modules/MSCloudLoginAssistant/Workloads/AdminAPI.ps1 [hunk 1]
```diff
@@ -6,33 +6,33 @@ function Connect-MSCloudLoginAdminAPI
     $WarningPreference = 'SilentlyContinue'
```
Comment (ykuijs) [inline Modules/MSCloudLoginAssistant/Workloads/AdminAPI.ps1:6]: Shouldn't we also remove this WarningPreference as well?

File: Modules/MSCloudLoginAssistant/Workloads/AdminAPI.ps1 [hunk 2]
```diff
@@ -151,20 +153,20 @@ function Connect-MSCloudLoginAdminAPIWithCertificateThumbprint
     Param()
     $WarningPreference = 'SilentlyContinue'
```
Comment (ykuijs) [inline Modules/MSCloudLoginAssistant/Workloads/AdminAPI.ps1:154]: Same comment about WarningPreference as earlier.

File: Modules/MSCloudLoginAssistant/Workloads/Azure.ps1 [hunk 3]
```diff
@@ -0,0 +1,113 @@
+function Connect-MSCloudLoginAzure
+{
+    [CmdletBinding()]
+    param()
+
+    $WarningPreference = 'SilentlyContinue'
```
Comment (ykuijs) [inline Modules/MSCloudLoginAssistant/Workloads/Azure.ps1:6]: Same comment about WarningPreference as earlier.

File: Modules/MSCloudLoginAssistant/Workloads/AzureDevOPS.ps1 [hunk 4]
```diff
@@ -6,32 +6,32 @@ function Connect-MSCloudLoginAzureDevOPS
     $WarningPreference = 'SilentlyContinue'
```
Comment (ykuijs) [inline Modules/MSCloudLoginAssistant/Workloads/AzureDevOPS.ps1:6]: Same comment about WarningPreference as earlier.

File: Modules/MSCloudLoginAssistant/Workloads/AzureDevOPS.ps1 [hunk 5]
```diff
@@ -153,21 +155,21 @@ function Connect-MSCloudLoginAzureDevOPSWithCertificateThumbprint
     Param()
     $WarningPreference = 'SilentlyContinue'
```
Comment (ykuijs) [inline Modules/MSCloudLoginAssistant/Workloads/AzureDevOPS.ps1:156]: Same comment about WarningPreference as earlier.

File: Modules/MSCloudLoginAssistant/Workloads/DefenderForEndpoint.ps1 [hunk 6]
```diff
@@ -5,32 +5,32 @@ function Connect-MSCloudLoginDefenderForEndpoint
 
     $ProgressPreference = 'SilentlyContinue'
     $WarningPreference = 'SilentlyContinue'
```
Comment (ykuijs) [inline Modules/MSCloudLoginAssistant/Workloads/DefenderForEndpoint.ps1:7]: Same comment about WarningPreference as earlier.

File: Modules/MSCloudLoginAssistant/Workloads/DefenderForEndpoint.ps1 [hunk 7]
```diff
@@ -159,29 +161,30 @@ function Connect-MSCloudLoginDefenderForEndpointWithAppSecret
...
+    $Script:MSCloudLoginConnectionProfile.DefenderForEndpoint.AccessToken = $managementToken.token_type.ToString() + ' ' + $managementToken.access_token.ToString()
 }
 
 function Connect-MSCloudLoginDefenderForEndpointWithCertificateThumbprint
 {
     [CmdletBinding()]
-    Param()
+    param()
+
     $WarningPreference = 'SilentlyContinue'
```
Comment (ykuijs) [inline Modules/MSCloudLoginAssistant/Workloads/DefenderForEndpoint.ps1:172]: Same comment about WarningPreference as earlier.

**Review by ykuijs — COMMENTED** [review]
There are several instances where you left the WarningPreference variable being changed to SilentlyContinue, but Write-Warning is being used. I marked a few of them, but not all.

Was there a reason for leaving these in there?

**Comment by FabienTschanz** [comment]
@ykuijs Not really a reason - I simply left it as it was and didn't want to change it. No problem with enabling it though, can easily do that. 
One question: I currently have a check on the `Microsoft.Graph.Identity.DirectoryManagement` because of the `Get-MgDomain` cmdlet. If this module is not installed, it will throw on calling. Do we actually want that (Microsoft365DSC does not require it) or do we ignore it? Or would it be better to simply not use `Get-MgDomain` in the module and make it independent?

File: Modules/MSCloudLoginAssistant/MSCloudLoginAssistant.psm1 [hunk 8]
```diff
@@ -263,74 +284,202 @@ function Connect-M365Tenant
...
+    This function resets the entire connection profile.
+.DESCRIPTION
+    This function resets the entire connection profile. It is used to disconnect all workloads and reset the connection profile.
+.EXAMPLE
+    Reset-MSCloudLoginConnectionProfileContext
+#>
+function Reset-MSCloudLoginConnectionProfileContext
+{
+    $source = 'Reset-MSCloudLoginConnectionProfileContext'
+    Add-MSCloudLoginAssistantEvent -Message 'Resetting connection profile' -Source $source
```
Comment (ykuijs) [inline Modules/MSCloudLoginAssistant/MSCloudLoginAssistant.psm1:395]: Shouldn't we add some logic to disconnect any open sessions? That way we clean up everything that we have created. Is also one of the Exchange best practices.

Pseudo code:
If ConnectedToExchange Then DisconnectExchange
If ConnectedToGraph Then DisconnectGraph
Etc, etc

**Comment by ykuijs** [comment]
> @ykuijs Not really a reason - I simply left it as it was and didn't want to change it. No problem with enabling it though, can easily do that. 

Please remove. I prefer removing all blocking of output that may come in useful in troubleshooting.

> One question: I currently have a check on the `Microsoft.Graph.Identity.DirectoryManagement` because of the `Get-MgDomain` cmdlet. If this module is not installed, it will throw on calling. Do we actually want that (Microsoft365DSC does not require it) or do we ignore it? Or would it be better to simply not use `Get-MgDomain` in the module and make it independent?

M365DSC doesn't require any modules via the module manifest, but does have prerequisites that you have to install via the Update-M365DSCDependencies cmdlet. That module is part of the prerequisites of M365DSC. So when you install all M365DSC prereqs, this module will also be installed. However, if you are using the MSCloudLoginAssistant for something else (standalone), you can run into issues because that module is not installed. So from that perspective, I like to have this check in.

**Review by ykuijs — APPROVED** [review]
LGTM

**Comment by ykuijs** [comment]
@NikCharlebois, this PR can be merged after PR #185

**Pull request merged**